File: VB/XPO/CustomEventsAndResources.Module/BusinessObjects/CustomEventWithUserResources.vb
Imports DevExpress.ExpressApp.Filtering
Imports DevExpress.ExpressApp.Model
Imports DevExpress.ExpressApp.SystemModule.Notifications
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.Base.General
Imports DevExpress.Persistent.Base.General.Compatibility
Imports DevExpress.Persistent.Validation
Imports DevExpress.XtraScheduler.Xml
Imports System.ComponentModel
Imports System.Xml
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Xpo

Namespace CustomEventsAndResources.Module.BusinessObjects

    <DefaultClassOptions>
    Public Class CustomEventWithUserResources
        Inherits BaseObject
        Implements IEvent, IRecurrentEvent, IReminderEvent

#Region "Base Properties"
        <Browsable(False)>
        Public ReadOnly Property AppointmentId As Object Implements IEvent.AppointmentId
            Get
                Return Me.Oid
            End Get
        End Property

        Public Property Subject As String Implements IEvent.Subject
            Get
                Return GetPropertyValue(Of String)(NameOf(CustomEventWithUserResources.Subject))
            End Get

            Set(ByVal value As String)
                SetPropertyValue(NameOf(CustomEventWithUserResources.Subject), value)
            End Set
        End Property

        Public Property Description As String Implements IEvent.Description
            Get
                Return GetPropertyValue(Of String)(NameOf(CustomEventWithUserResources.Description))
            End Get

            Set(ByVal value As String)
                SetPropertyValue(NameOf(CustomEventWithUserResources.Description), value)
            End Set
        End Property

        <Indexed>
        Public Property StartOn As DateTime
            Get
                Return GetPropertyValue(Of DateTime)(NameOf(CustomEventWithUserResources.StartOn))
            End Get

            Set(ByVal value As DateTime)
                SetPropertyValue(NameOf(CustomEventWithUserResources.StartOn), value)
            End Set
        End Property

        <Indexed>
        Public Property EndOn As DateTime
            Get
                Return GetPropertyValue(Of DateTime)(NameOf(CustomEventWithUserResources.EndOn))
            End Get

            Set(ByVal value As DateTime)
                SetPropertyValue(NameOf(CustomEventWithUserResources.EndOn), value)
            End Set
        End Property

        <ImmediatePostData>
        Public Property AllDay As Boolean Implements IEvent.AllDay
            Get
                Return GetPropertyValue(Of Boolean)(NameOf(CustomEventWithUserResources.AllDay))
            End Get

            Set(ByVal value As Boolean)
                SetPropertyValue(NameOf(CustomEventWithUserResources.AllDay), value)
            End Set
        End Property

        Public Property Location As String Implements IEvent.Location
            Get
                Return GetPropertyValue(Of String)(NameOf(CustomEventWithUserResources.Location))
            End Get

            Set(ByVal value As String)
                SetPropertyValue(NameOf(CustomEventWithUserResources.Location), value)
            End Set
        End Property

        Public Property Label As Integer Implements IEvent.Label
            Get
                Return GetPropertyValue(Of Integer)(NameOf(CustomEventWithUserResources.Label))
            End Get

            Set(ByVal value As Integer)
                SetPropertyValue(NameOf(CustomEventWithUserResources.Label), value)
            End Set
        End Property

        Public Property Status As Integer Implements IEvent.Status
            Get
                Return GetPropertyValue(Of Integer)(NameOf(CustomEventWithUserResources.Status))
            End Get

            Set(ByVal value As Integer)
                SetPropertyValue(NameOf(CustomEventWithUserResources.Status), value)
            End Set
        End Property

        <Browsable(False)>
        Public Property Type As Integer Implements IEvent.Type
            Get
                Return GetPropertyValue(Of Integer)(NameOf(CustomEventWithUserResources.Type))
            End Get

            Set(ByVal value As Integer)
                SetPropertyValue(NameOf(CustomEventWithUserResources.Type), value)
            End Set
        End Property

#End Region
#Region "Resources"
        <Persistent("ResourceIds"), Size(SizeAttribute.Unlimited), ObjectValidatorIgnoreIssue(GetType(ObjectValidatorLargeNonDelayedMember))>
        Private resourceIds As [String]

        Private Sub UpdateResources()
            Resources.SuspendChangedEvents()
            Try
                While Resources.Count > 0
                    Resources.Remove(Resources(0))
                End While

                If Not [String].IsNullOrEmpty(resourceIds) Then
                    Dim xmlDocument As XmlDocument = DevExpress.Utils.SafeXml.CreateDocument(resourceIds)
                    For Each xmlNode As XmlNode In xmlDocument.DocumentElement.ChildNodes
                        Dim loader As AppointmentResourceIdXmlLoader = New AppointmentResourceIdXmlLoader(xmlNode)
                        Dim keyMemberValue As [Object] = loader.ObjectFromXml()
                        Dim resource As ApplicationUser = Session.GetObjectByKey(Of ApplicationUser)(New Guid(keyMemberValue.ToString()))
                        If resource IsNot Nothing Then
                            Resources.Add(resource)
                        End If
                    Next
                End If
            Finally
                Resources.ResumeChangedEvents()
            End Try
        End Sub

        Private Sub Resources_ListChanged(ByVal sender As Object, ByVal e As ListChangedEventArgs)
            If e.ListChangedType Is ListChangedType.ItemAdded OrElse e.ListChangedType Is ListChangedType.ItemDeleted Then
                UpdateResourceIds()
                OnChanged(NameOf(CustomEventWithUserResources.ResourceId))
            End If
        End Sub

        Protected Overrides Sub OnLoaded()
            MyBase.OnLoaded()
            ' B136420, Q255811
            If Resources.IsLoaded AndAlso Not Session.IsNewObject(Me) Then
                Resources.Reload()
            End If
        End Sub

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
            Resources.ListChanged += New ListChangedEventHandler(AddressOf Resources_ListChanged)
        End Sub

        Public Sub UpdateResourceIds()
            resourceIds = "<ResourceIds>" & Microsoft.VisualBasic.Constants.vbCrLf
            For Each resource As ApplicationUser In Resources
                resourceIds += String.Format("<ResourceId Type=""{0}"" Value=""{1}"" />" & Microsoft.VisualBasic.Constants.vbCrLf, resource.Id.GetType().FullName, resource.Id)
            Next

            resourceIds += "</ResourceIds>"
        End Sub

        <Association("CustomEventWithUserResources-ApplicationUser")>
        Public ReadOnly Property Resources As XPCollection(Of ApplicationUser)
            Get
                Return GetCollection(Of ApplicationUser)(NameOf(CustomEventWithUserResources.Resources))
            End Get
        End Property

        <NonPersistent(), Browsable(False)>
        Public Property ResourceId As [String]
            Get
                If resourceIds Is Nothing Then
                    UpdateResourceIds()
                End If

                Return resourceIds
            End Get

            Set(ByVal value As [String])
                If resourceIds IsNot value Then
                    resourceIds = value
                    UpdateResources()
                End If
            End Set
        End Property

#End Region
#Region "IRecurrentEvent"
        <Persistent("RecurrencePattern")>
        Private recurrencePatternField As [Event]

        Private recurrenceInfoXmlField As String

        <NonCloneable>
        <DevExpress.Xpo.DisplayName("Recurrence"), Size(SizeAttribute.Unlimited), ObjectValidatorIgnoreIssue(GetType(ObjectValidatorLargeNonDelayedMember))>
        Public Property RecurrenceInfoXml As String Implements IRecurrentEvent.RecurrenceInfoXml
            Get
                Return recurrenceInfoXmlField
            End Get

            Set(ByVal value As String)
                SetPropertyValue(NameOf(CustomEventWithUserResources.RecurrenceInfoXml), recurrenceInfoXmlField, value)
            End Set
        End Property

        <PersistentAlias(NameOf(recurrencePatternField))>
        Public Property RecurrencePattern As IRecurrentEvent Implements IRecurrentEvent.RecurrencePattern
            Get
                Return recurrencePatternField
            End Get

            Set(ByVal value As IRecurrentEvent)
                SetPropertyValue(NameOf(CustomEventWithUserResources.RecurrencePattern), recurrencePatternField, CType(value, [Event]))
            End Set
        End Property

#End Region
#Region "IReminderEvent"
        Private reminderInfoXmlField As String

        Private remindInField As TimeSpan?

        Private alarmTimeField As Nullable(Of DateTime)

        Private postponeTimes As IList(Of PostponeTime)

        Private isPostponedField As Boolean

        Private Sub UpdateRemindersInfoXml(ByVal UpdateAlarmTime As Boolean)
            If RemindIn.HasValue AndAlso AlarmTime.HasValue Then
                Dim aptReminder As AppointmentReminderInfo = New AppointmentReminderInfo()
                Dim reminderInfo As ReminderInfo = New ReminderInfo()
                reminderInfo.TimeBeforeStart = RemindIn.Value
                If UpdateAlarmTime Then
                    reminderInfo.AlertTime = AlarmTime.Value
                Else
                    reminderInfo.AlertTime = StartOn - RemindIn.Value
                End If

                aptReminder.ReminderInfos.Add(reminderInfo)
                reminderInfoXmlField = aptReminder.ToXml()
            Else
                reminderInfoXmlField = Nothing
            End If
        End Sub

        Private Sub UpdateAlarmTime()
            If Not String.IsNullOrEmpty(reminderInfoXmlField) Then
                Dim appointmentReminderInfo As AppointmentReminderInfo = New AppointmentReminderInfo()
                Try
                    appointmentReminderInfo.FromXml(reminderInfoXmlField)
                    Me.alarmTimeField = appointmentReminderInfo.ReminderInfos(CInt(0)).AlertTime
                Catch e As XmlException
                    Tracing.Tracer.LogError(e)
                End Try
            Else
                alarmTimeField = Nothing
                remindInField = Nothing
                IsPostponed = False
            End If
        End Sub

        Private Function CreatePostponeTimes() As IList(Of PostponeTime)
            Dim result As IList(Of PostponeTime) = PostponeTime.CreateDefaultPostponeTimesList()
            result.Add(New PostponeTime("None", Nothing, "None"))
            result.Add(New PostponeTime("AtStartTime", TimeSpan.Zero, "0 minutes"))
            Dim args As CustomizeNotificationsPostponeTimeListEventArgs = New CustomizeNotificationsPostponeTimeListEventArgs(result)
            RaiseEvent CustomizeReminderTimeLookupEvent(Me, args)
            PostponeTime.SortPostponeTimesList(args.PostponeTimesList)
            Return args.PostponeTimesList
        End Function

        <ImmediatePostData>
        <NonPersistent>
        <ModelDefault("AllowClear", "False")>
        <DataSourceProperty(NameOf(CustomEventWithUserResources.PostponeTimeList))>
        <SearchMemberOptions(SearchMemberMode.Exclude)>
        <VisibleInDetailView(False), VisibleInListView(False)>
        Public Property ReminderTime As PostponeTime
            Get
                If RemindIn.HasValue Then
                    Return PostponeTimeList.Where(Function(x) x.RemindIn IsNot Nothing AndAlso x.RemindIn.Value Is remindInField.Value).FirstOrDefault()
                Else
                    Return PostponeTimeList.Where(Function(x) x.RemindIn Is Nothing).FirstOrDefault()
                End If
            End Get

            Set(ByVal value As PostponeTime)
                If Not IsLoading AndAlso value IsNot Nothing Then
                    If value.RemindIn.HasValue Then
                        RemindIn = value.RemindIn.Value
                    Else
                        RemindIn = Nothing
                    End If
                End If
            End Set
        End Property

        <Browsable(False)>
        Public ReadOnly Property PostponeTimeList As IEnumerable(Of PostponeTime)
            Get
                If postponeTimes Is Nothing Then
                    postponeTimes = CreatePostponeTimes()
                End If

                Return postponeTimes
            End Get
        End Property

        <Browsable(False)>
        Public Property RemindIn As TimeSpan?
            Get
                Return remindInField
            End Get

            Set(ByVal value As TimeSpan?)
                SetPropertyValue(NameOf(CustomEventWithUserResources.RemindIn), remindInField, value)
            End Set
        End Property

        <NonCloneable>
        <Browsable(False)>
        <Size(200)>
        Public Property ReminderInfoXml As String Implements IReminderEvent.ReminderInfoXml
            Get
                Return reminderInfoXmlField
            End Get

            Set(ByVal value As String)
                SetPropertyValue(NameOf(CustomEventWithUserResources.ReminderInfoXml), reminderInfoXmlField, value)
                If Not IsLoading Then
                    UpdateAlarmTime()
                End If
            End Set
        End Property

        <VisibleInDetailView(False), VisibleInListView(False), VisibleInLookupListView(False)>
        Public ReadOnly Property NotificationMessage As String Implements ISupportNotifications.NotificationMessage
            Get
                Return Subject
            End Get
        End Property

        <Browsable(False)>
        Public ReadOnly Property UniqueId As Object Implements ISupportNotifications.UniqueId
            Get
                Return Me.Oid
            End Get
        End Property

        <VisibleInDetailView(False), VisibleInListView(False), VisibleInLookupListView(False)>
        Public Property AlarmTime As DateTime?
            Get
                Return alarmTimeField
            End Get

            Set(ByVal value As DateTime?)
                SetPropertyValue(NameOf(CustomEventWithUserResources.AlarmTime), alarmTimeField, value)
                If Not IsLoading Then
                    If value Is Nothing Then
                        remindInField = Nothing
                        IsPostponed = False
                    End If

                    UpdateRemindersInfoXml(True)
                End If
            End Set
        End Property

        <VisibleInDetailView(False), VisibleInListView(False), VisibleInLookupListView(False)>
        Public Property IsPostponed As Boolean Implements ISupportNotifications.IsPostponed
            Get
                Return isPostponedField
            End Get

            Set(ByVal value As Boolean)
                SetPropertyValue(NameOf(CustomEventWithUserResources.IsPostponed), isPostponedField, value)
            End Set
        End Property

        Public Event CustomizeReminderTimeLookup As EventHandler(Of CustomizeNotificationsPostponeTimeListEventArgs)

#End Region
#Region "Blazor compatibility"
        <NonPersistent(), Browsable(False)>
        Public Property ResourceIdBlazor As Object
            Get
                If Resources.Count = 1 Then
                    Return Resources(0).Id
                End If

                Return Nothing
            End Get

            Set(ByVal value As Object)
                Resources.SuspendChangedEvents()
                Try
                    While Resources.Count > 0
                        Resources.Remove(Resources(0))
                    End While

                    If value IsNot Nothing Then
                        Resources.Add(Session.GetObjectByKey(Of ApplicationUser)(value))
                    End If
                Finally
                    Resources.ResumeChangedEvents()
                End Try
            End Set
        End Property

        <NonPersistent(), Browsable(False)>
        Public Property RecurrenceInfoXmlBlazor As String
            Get
                Return RecurrenceInfoXml?.ToNewRecurrenceInfoXml()
            End Get

            Set(ByVal value As String)
                RecurrenceInfoXml = value?.ToOldRecurrenceInfoXml()
            End Set
        End Property

#End Region
#Region "Construction"
        Public Overrides Sub AfterConstruction()
            MyBase.AfterConstruction()
            StartOn = DateTime.Now
            EndOn = StartOn.AddHours(1)
        End Sub

#End Region
#Region "Validation"
        <Browsable(False)>
        <RuleFromBoolProperty("CustomEventWithUserResourcesIntervalValid", DefaultContexts.Save, "The start date must be less than the end date", SkipNullOrEmptyValues:=False, UsedProperties:="StartOn, EndOn")>
        Public ReadOnly Property IsIntervalValid As Boolean
            Get
                Return StartOn <= EndOn
            End Get
        End Property
#End Region
    End Class
End Namespace

File: CS/XPO/CustomEventsAndResources.Module/BusinessObjects/CustomEventWithUserResources.cs
using DevExpress.ExpressApp.Filtering;
using DevExpress.ExpressApp.Model;
using DevExpress.ExpressApp.SystemModule.Notifications;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.Base.General;
using DevExpress.Persistent.Base.General.Compatibility;
using DevExpress.Persistent.Validation;
using DevExpress.XtraScheduler.Xml;
using System.ComponentModel;
using System.Xml;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Xpo;

namespace CustomEventsAndResources.Module.BusinessObjects;

[DefaultClassOptions]
public class CustomEventWithUserResources : BaseObject, IEvent, IRecurrentEvent, IReminderEvent
{
    #region Base Properties
    [Browsable(false)]
    public object AppointmentId {
        get => Oid;
    }
    public string Subject {
        get => GetPropertyValue<string>(nameof(Subject));
        set => SetPropertyValue(nameof(Subject), value);
    }
    public string Description {
        get => GetPropertyValue<string>(nameof(Description));
        set => SetPropertyValue(nameof(Description), value);
    }
    [Indexed]
    public DateTime StartOn {
        get => GetPropertyValue<DateTime>(nameof(StartOn));
        set => SetPropertyValue(nameof(StartOn), value);
    }
    [Indexed]
    public DateTime EndOn {
        get => GetPropertyValue<DateTime>(nameof(EndOn));
        set => SetPropertyValue(nameof(EndOn), value);
    }
    [ImmediatePostData]
    public bool AllDay {
        get => GetPropertyValue<bool>(nameof(AllDay));
        set => SetPropertyValue(nameof(AllDay), value);
    }
    public string Location {
        get => GetPropertyValue<string>(nameof(Location));
        set => SetPropertyValue(nameof(Location), value);
    }
    public int Label {
        get => GetPropertyValue<int>(nameof(Label));
        set => SetPropertyValue(nameof(Label), value);
    }
    public int Status {
        get => GetPropertyValue<int>(nameof(Status));
        set => SetPropertyValue(nameof(Status), value);
    }
    [Browsable(false)]
    public int Type {
        get => GetPropertyValue<int>(nameof(Type));
        set => SetPropertyValue(nameof(Type), value);
    }
    #endregion
    #region Resources
    [Persistent("ResourceIds"), Size(SizeAttribute.Unlimited), ObjectValidatorIgnoreIssue(typeof(ObjectValidatorLargeNonDelayedMember))]
    private String resourceIds;
    private void UpdateResources() {
        Resources.SuspendChangedEvents();
        try {
            while (Resources.Count > 0) {
                Resources.Remove(Resources[0]);
            }
            if (!String.IsNullOrEmpty(resourceIds)) {
                XmlDocument xmlDocument = DevExpress.Utils.SafeXml.CreateDocument(resourceIds);
                foreach (XmlNode xmlNode in xmlDocument.DocumentElement.ChildNodes) {
                    AppointmentResourceIdXmlLoader loader = new AppointmentResourceIdXmlLoader(xmlNode);
                    Object keyMemberValue = loader.ObjectFromXml();
                    ApplicationUser resource = Session.GetObjectByKey<ApplicationUser>(new Guid(keyMemberValue.ToString()));
                    if (resource != null) {
                        Resources.Add(resource);
                    }
                }
            }
        } finally {
            Resources.ResumeChangedEvents();
        }
    }
    private void Resources_ListChanged(object sender, ListChangedEventArgs e) {
        if ((e.ListChangedType == ListChangedType.ItemAdded) ||
            (e.ListChangedType == ListChangedType.ItemDeleted)) {
            UpdateResourceIds();
            OnChanged(nameof(ResourceId));
        }
    }

    protected override void OnLoaded() {
        base.OnLoaded();
        // B136420, Q255811
        if (Resources.IsLoaded && !Session.IsNewObject(this)) {
            Resources.Reload();
        }
    }

    public CustomEventWithUserResources(Session session)
        : base(session) {
        Resources.ListChanged += new ListChangedEventHandler(Resources_ListChanged);
    }
    public void UpdateResourceIds() {
        resourceIds = "<ResourceIds>\r\n";
        foreach (ApplicationUser resource in Resources) {
            resourceIds += string.Format("<ResourceId Type=\"{0}\" Value=\"{1}\" />\r\n", resource.Id.GetType().FullName, resource.Id);
        }
        resourceIds += "</ResourceIds>";
    }

    [Association("CustomEventWithUserResources-ApplicationUser")]
    public XPCollection<ApplicationUser> Resources {
        get { return GetCollection<ApplicationUser>(nameof(Resources)); }
    }
    [NonPersistent(), Browsable(false)]
    public String ResourceId {
        get {
            if (resourceIds == null) {
                UpdateResourceIds();
            }
            return resourceIds;
        }
        set {
            if (resourceIds != value) {
                resourceIds = value;
                UpdateResources();
            }
        }
    }
    #endregion
    #region IRecurrentEvent
    [Persistent("RecurrencePattern")]
    private Event recurrencePattern;
    private string recurrenceInfoXml;
    [NonCloneable]
    [DevExpress.Xpo.DisplayName("Recurrence"), Size(SizeAttribute.Unlimited), ObjectValidatorIgnoreIssue(typeof(ObjectValidatorLargeNonDelayedMember))]
    public string RecurrenceInfoXml {
        get { return recurrenceInfoXml; }
        set { SetPropertyValue(nameof(RecurrenceInfoXml), ref recurrenceInfoXml, value); }
    }
    [PersistentAlias(nameof(recurrencePattern))]
    public IRecurrentEvent RecurrencePattern {
        get { return recurrencePattern; }
        set { SetPropertyValue(nameof(RecurrencePattern), ref recurrencePattern, (Event)value); }
    }
    #endregion
    #region IReminderEvent

    private string reminderInfoXml;
    private TimeSpan? remindIn;
    private Nullable<DateTime> alarmTime;
    private IList<PostponeTime> postponeTimes;
    private bool isPostponed;

    private void UpdateRemindersInfoXml(bool UpdateAlarmTime) {
        if (RemindIn.HasValue && AlarmTime.HasValue) {
            AppointmentReminderInfo aptReminder = new AppointmentReminderInfo();
            ReminderInfo reminderInfo = new ReminderInfo();
            reminderInfo.TimeBeforeStart = RemindIn.Value;
            if (UpdateAlarmTime)
                reminderInfo.AlertTime = AlarmTime.Value;
            else
                reminderInfo.AlertTime = StartOn - RemindIn.Value;
            aptReminder.ReminderInfos.Add(reminderInfo);
            reminderInfoXml = aptReminder.ToXml();
        } else {
            reminderInfoXml = null;
        }
    }
    private void UpdateAlarmTime() {
        if (!string.IsNullOrEmpty(reminderInfoXml)) {
            AppointmentReminderInfo appointmentReminderInfo = new AppointmentReminderInfo();
            try {
                appointmentReminderInfo.FromXml(reminderInfoXml);
                alarmTime = appointmentReminderInfo.ReminderInfos[0].AlertTime;
            } catch (XmlException e) {
                Tracing.Tracer.LogError(e);
            }
        } else {
            alarmTime = null;
            remindIn = null;
            IsPostponed = false;
        }
    }
    private IList<PostponeTime> CreatePostponeTimes() {
        IList<PostponeTime> result = PostponeTime.CreateDefaultPostponeTimesList();
        result.Add(new PostponeTime("None", null, "None"));
        result.Add(new PostponeTime("AtStartTime", TimeSpan.Zero, "0 minutes"));
        CustomizeNotificationsPostponeTimeListEventArgs args = new CustomizeNotificationsPostponeTimeListEventArgs(result);
        if (CustomizeReminderTimeLookup != null) {
            CustomizeReminderTimeLookup(this, args);
        }
        PostponeTime.SortPostponeTimesList(args.PostponeTimesList);
        return args.PostponeTimesList;
    }
    [ImmediatePostData]
    [NonPersistent]
    [ModelDefault("AllowClear", "False")]
    [DataSourceProperty(nameof(PostponeTimeList))]
    [SearchMemberOptions(SearchMemberMode.Exclude)]
    [VisibleInDetailView(false), VisibleInListView(false)]
    public PostponeTime ReminderTime {
        get {
            if (RemindIn.HasValue) {
                return PostponeTimeList.Where(x => (x.RemindIn != null && x.RemindIn.Value == remindIn.Value)).FirstOrDefault();
            } else {
                return PostponeTimeList.Where(x => x.RemindIn == null).FirstOrDefault();
            }
        }
        set {
            if (!IsLoading && (value != null)) {
                if (value.RemindIn.HasValue) {
                    RemindIn = value.RemindIn.Value;
                } else {
                    RemindIn = null;
                }
            }
        }
    }
    [Browsable(false)]
    public IEnumerable<PostponeTime> PostponeTimeList {
        get {
            if (postponeTimes == null) {
                postponeTimes = CreatePostponeTimes();
            }
            return postponeTimes;
        }
    }
    [Browsable(false)]
    public TimeSpan? RemindIn {
        get { return remindIn; }
        set {
            SetPropertyValue(nameof(RemindIn), ref remindIn, value);
        }
    }
    [NonCloneable]
    [Browsable(false)]
    [Size(200)]
    public string ReminderInfoXml {
        get { return reminderInfoXml; }
        set {
            SetPropertyValue(nameof(ReminderInfoXml), ref reminderInfoXml, value);
            if (!IsLoading) {
                UpdateAlarmTime();
            }
        }
    }
    [VisibleInDetailView(false), VisibleInListView(false), VisibleInLookupListView(false)]
    public string NotificationMessage {
        get { return Subject; }
    }
    [Browsable(false)]
    public object UniqueId {
        get { return Oid; }
    }
    [VisibleInDetailView(false), VisibleInListView(false), VisibleInLookupListView(false)]
    public DateTime? AlarmTime {
        get { return alarmTime; }
        set {
            SetPropertyValue(nameof(AlarmTime), ref alarmTime, value);
            if (!IsLoading) {
                if (value == null) {
                    remindIn = null;
                    IsPostponed = false;
                }
                UpdateRemindersInfoXml(true);
            }
        }
    }
    [VisibleInDetailView(false), VisibleInListView(false), VisibleInLookupListView(false)]
    public bool IsPostponed {
        get { return isPostponed; }
        set { SetPropertyValue(nameof(IsPostponed), ref isPostponed, value); }
    }
    public event EventHandler<CustomizeNotificationsPostponeTimeListEventArgs> CustomizeReminderTimeLookup;
    #endregion
    #region Blazor compatibility
    [NonPersistent(), Browsable(false)]
    public string RecurrenceInfoXmlBlazor {
        get { return RecurrenceInfoXml?.ToNewRecurrenceInfoXml(); }
        set { RecurrenceInfoXml = value?.ToOldRecurrenceInfoXml(); }
    }
    #endregion
    #region Construction
    public override void AfterConstruction() {
        base.AfterConstruction();
        StartOn = DateTime.Now;
        EndOn = StartOn.AddHours(1);
    }
    #endregion
    #region Validation
    [Browsable(false)]
    [RuleFromBoolProperty("CustomEventWithUserResourcesIntervalValid", DefaultContexts.Save, "The start date must be less than the end date", SkipNullOrEmptyValues = false, UsedProperties = "StartOn, EndOn")]
    public bool IsIntervalValid { get { return StartOn <= EndOn; } }
    #endregion
}
